### ISSUE TITLE
Add common fields

### ISSUE THREAD

**Issue by bob606** [birchill/discriminator#1]
Hello, I have the following problem. 

I have type which follows the discriminator pattern. But the type has additional common fields `id, title`
``` typescript
type ModuleBase ={
  id: string,
  title: string,
  type: ModulType,
};

type ChapterHeadingModule = ModuleBase & {
  type: ModulType.CHAPTER_HEADING,
  heading: string,
};
```
Now I thought I can combine the two schemas with `assign` like so:
``` typescript
const moduleBaseSchema: Describe<Omit<ModuleBase, 'type'>> = object({
  id: string(),
  title: string(),
});

const moduleDiscriminatorSchema = discriminator('type', {
  CHAPTER_HEADING: object({
    heading: string(),
  }),
});

const postModuleInputSchema = assign(moduleBaseSchema, moduleDiscriminatorSchema);
```
But the discriminator schema doesn't have the right shape. If I call `validate` is an error occurring: `TypeError: struct.validator is not a function or its return value is not iterable`.

Any ideas hot solve this :)?

**Comment by birtles** [comment]
Hi!

Thanks for the issue report. I tried to reproduce it but I couldn't quite understand the desired effect.

Perhaps you need to use `assign` inside the `discriminator` branch something like the following?

```ts
import * as s from 'superstruct';
import { discriminator } from '@birchill/discriminator';

type ModulType = 'CHAPTER_HEADING' | 'OTHER';

type ModuleBase ={
  id: string,
  title: string,
  type: ModulType,
};

type ChapterHeadingModule = ModuleBase & {
  type: 'CHAPTER_HEADING',
  heading: string,
};

const moduleBaseSchema: s.Describe<Omit<ModuleBase, 'type'>> = s.object({
  id: s.string(),
  title: s.string(),
});

const moduleDiscriminatorSchema = discriminator('type', {
  CHAPTER_HEADING: s.assign(moduleBaseSchema, s.object({
    heading: s.string(),
  })),
  OTHER: moduleBaseSchema
});
```

**Comment by birtles** [comment]
For my reference, here is a sandbox of the above: https://codesandbox.io/s/elated-moser-1eyxrw?file=/src/index.ts

**Comment by bob606** [comment]
Hey thank you for your fast reply, I appreciate it a lot. I tried to make the example less complex and watered it down too much in the process. 

The Type I actually want to validate looks more or less like this, just more types in the union (the types in the union are modeled after the discriminator pattern: 
*** typescript
type Module = (
  ChapterHeadingModule |
  EditorialModule |
  PlaceholderModule
) & ModuleBase;
***
I love your solution for the fact that it works.

In general I just would love to know, if there is a way that I don't have to write `assign(moduleBaseSchema` into every discriminator case. 

Also is there a way to use `Describe` with the `discriminator()` function or how are you solving the type safety issue :)?

**Comment by birtles** [comment]
> Hey thank you for your fast reply, I appreciate it a lot. I tried to make the example less complex and watered it down too much in the process.

Not at all! Thanks for trying out the discriminator utility.

> The Type I actually want to validate looks more or less like this, just more types in the union (the types in the union are modeled after the discriminator pattern:
> 
> ```ts
> type Module = (
>   ChapterHeadingModule |
>   EditorialModule |
>   PlaceholderModule
> ) & ModuleBase;
> ```
> 
> I love your solution for the fact that it works. Just to have asked, is there a way that I don't have to write `assign(moduleBaseSchema` into every discriminator case?

I see, that makes sense. I'm not sure, but I think it might not currently be possible to have `assign()` accept a `discriminator` object since it always produces either a `type` or an `object`: https://github.com/ianstormtaylor/superstruct/blob/41d7fdd09a0c0f0291b03357e487420d4ece6b56/src/structs/utilities.ts#L55-L60

Ian suggested I make a PR to add this to `superstruct` over in https://github.com/ianstormtaylor/superstruct/issues/897#issuecomment-915449301 so once I do that, maybe we can make `assign` work with `discriminator`.

For what it's worth, I have often found that `Describe` doesn't work properly for many complicated cases. As a result, I often use `Infer` instead. As long as you are not exporting the types as part of a library interface, it might work for you.

For example:

```ts

import * as s from 'superstruct';
import { discriminator } from '@birchill/discriminator';

const moduleBaseSchema = s.object({
  id: s.string(),
  title: s.string(),
});

const moduleDiscriminatorSchema = discriminator('type', {
  'CHAPTER_HEADING': s.assign(moduleBaseSchema, s.object({
    heading: s.string(),
  })),
  'EDITORIAL': s.assign(moduleBaseSchema, s.object({
    // editorial-specific fields go here
  })),
  'PLACEHOLDER': s.assign(moduleBaseSchema, s.object({
    // placeholder-specific fields go here
  }))
});

type Module = s.Infer<typeof moduleDiscriminatorSchema>;
type ModuleType = Module['type'];
type ChaperHeadingModule = Extract<Module, { type: 'CHAPTER_HEADING' }>;
type EditorialModule = Extract<Module, { type: 'EDITORIAL' }>;
type PlaceholderModule = Extract<Module, { type: 'PLACEHOLDER' }>;
```

**Comment by birtles** [comment]
Alternatively, I often find that I don't need the `type` field in the different type definitions and so the following is easier:

```ts
import * as s from 'superstruct';
import { discriminator } from '@birchill/discriminator';

const moduleBaseSchema = s.object({
  id: s.string(),
  title: s.string(),
});

const chapterHeadingSchema = s.assign(moduleBaseSchema, s.object({
  heading: s.string(),
}));

const editorialSchema = s.assign(moduleBaseSchema, s.object({
  // editorial-specific fields go here
}));

const placeholderSchema = s.assign(moduleBaseSchema, s.object({
  // placeholder-specific fields go here
}));

const moduleDiscriminatorSchema = discriminator('type', {
  'CHAPTER_HEADING': chapterHeadingSchema,
  'EDITORIAL': editorialSchema,
  'PLACEHOLDER': placeholderSchema
});

type Module = s.Infer<typeof moduleDiscriminatorSchema>;
type ModuleType = Module['type'];

type ChaperHeadingModule = s.Infer<typeof chapterHeadingSchema>;
type EditorialModule = s.Infer<typeof editorialSchema>;
type PlaceholderModule = s.Infer<typeof placeholderSchema>;
```

**Comment by bob606** [comment]
Thank you so much. Your library literally saved us so much time over here at our company. 

I deleted the validation question again, as I found the inferred type which automatically comes with the discriminator function plenty useful. Thank you you nevertheless for your detailed answer. 

I think once this discriminator library gains more traction, it would actually be useful to do add the issue to superstruct  but for now I can happily report that I tested the validation and everything works perfectly with the multiple `assign` approach. Again thank you so much for your insights :heart:
